components/brave_wallet/common/f4_jumble.cc

```diff
@@ -0,0 +1,156 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/common/f4_jumble.h"
+
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <string>
+
+#include "base/containers/span.h"
+#include "base/notreached.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "brave/third_party/argon2/src/src/blake2/blake2.h"
+
+namespace brave_wallet {
+
+// Sizes for Blake2b are defined here https://zips.z.cash/zip-0316#solution
+const size_t kMinMessageSize = 48u;
+const size_t kMaxMessageSize = 4184368u;
+const size_t kLeftSize = 64u;
+
+std::array<uint8_t, 16> GetHPersonalizer(uint8_t i) {
+  return {85,  65, 95,  70,  52, 74, 117,
+          109, 98, 108, 101, 95, 72, static_cast<uint8_t>(i),
+          0,   0};
+}
+
+std::array<uint8_t, 16> GetGPersonalizer(uint8_t i, uint16_t j) {
+  return {85,
+          65,
+          95,
+          70,
+          52,
+          74,
+          117,
+          109,
+          98,
+          108,
+          101,
+          95,
+          71,
+          static_cast<uint8_t>(i),
+          static_cast<uint8_t>(j & 0xFF),
+          static_cast<uint8_t>(j >> 8)};
+}
+
+std::vector<uint8_t> blake2b(const std::vector<uint8_t>& payload,
+                             const std::array<uint8_t, 16>& personalizer,
+                             size_t len) {
+  blake2b_state blake_state = {};
+  blake2b_param params = {};
+
+  params.digest_length = len;
+  params.fanout = 1;
+  params.depth = 1;
+  if (personalizer.size() != sizeof(params.personal)) {
+    NOTREACHED_NORETURN();
```

Comment:
crashing here doesn't seem ideal?

Reply:
Initially this was an optional but was changed to CHECK behavior due 
https://github.com/brave/brave-core/pull/21460#discussion_r1436705525 s

Reply:
agreed - crashing here seems less than ideal. Especially since some of these unified addresses will be user controlled. Passing the error up so it can be handled elsewhere seems like a safer way to handle this. @supermassive does that stay in line with what you were hoping for out of your comment above as well?

Reply:
This particular line compares sizes of fixed-sized arrays. Worth checking with static_assert.
`blake2b_*` calls below fail on invalid inputs(nullptr args, zero length args, etc) which is not the case. 
Returning an optional here just makes code unnecessarily more complicated.

Reply:
I've added other testcases and also linked to librustzcash test vectors when available